File: Cpp/ImageProcessing/EdgeDetection.cpp
#include "EdgeDetection.h"
#include "ComFunc.h"
#include "ComInfo.h"

/// <summary>
/// コンストラクタ
/// </summary>
EdgeDetection::EdgeDetection()
{
}

/// <summary>
/// デスクトラクタ
/// </summary>
EdgeDetection::~EdgeDetection()
{
}

/// <summary>
/// エッジ検出の実行
/// </summary>
/// <param name="bitmap">ビットマップ</param>
/// <param name="token">キャンセルトークン</param>
/// <returns>実行結果 成功/失敗</returns>
bool EdgeDetection::ImageProcessing(Bitmap^ bitmap, CancellationToken^ token)
{
	bool result = true;

	short mask[3][3] =
	{
		{ 1,  1, 1 },
		{ 1, -8, 1 },
		{ 1,  1, 1 }
	};

	int widthSize = bitmap->Width;
	int heightSize = bitmap->Height;
	int maskSize = 3;
	
	auto bitmapData = bitmap->LockBits(System::Drawing::Rectangle(0, 0, widthSize, heightSize), ImageLockMode::ReadWrite, PixelFormat::Format32bppArgb);

	int indexWidth;
	int indexHeight;

	for (indexHeight = 0; indexHeight < heightSize; indexHeight++)
	{
		if (token->IsCancellationRequested)
		{
			result = false;
			break;
		}

		for (indexWidth = 0; indexWidth < widthSize; indexWidth++)
		{
			if (token->IsCancellationRequested)
			{
				result = false;
				break;
			}

			Byte* pixel = (Byte*)bitmapData->Scan0.ToPointer() + indexHeight * bitmapData->Stride + indexWidth * 4;

			long totalBlue = 0;
			long totalGreen = 0;
			long totalRed = 0;
			int indexWidthMask;
			int indexHightMask;

			for (indexHightMask = 0; indexHightMask < maskSize; indexHightMask++)
			{
				for (indexWidthMask = 0; indexWidthMask < maskSize; indexWidthMask++)
				{
					if (indexWidth + indexWidthMask > 0 &&
						indexWidth + indexWidthMask < widthSize &&
						indexHeight + indexHightMask > 0 &&
						indexHeight + indexHightMask < heightSize)
					{
						Byte* pixelMaskArea = (Byte*)bitmapData->Scan0.ToPointer() + (indexHeight + indexHightMask) * bitmapData->Stride + (indexWidth + indexWidthMask) * 4;

						totalBlue += pixelMaskArea[ComInfo::Pixel::Type::B] * mask[indexWidthMask][indexHightMask];
						totalGreen += pixelMaskArea[ComInfo::Pixel::Type::G] * mask[indexWidthMask][indexHightMask];
						totalRed += pixelMaskArea[ComInfo::Pixel::Type::R] * mask[indexWidthMask][indexHightMask];
					}
				}
			}
			pixel[ComInfo::Pixel::Type::B] = ComFunc::LongToByte(totalBlue);
			pixel[ComInfo::Pixel::Type::G] = ComFunc::LongToByte(totalGreen);
			pixel[ComInfo::Pixel::Type::R] = ComFunc::LongToByte(totalRed);
		}
	}
	bitmap->UnlockBits(bitmapData);

	return result;
}

File: Vb/ImageProcessing/ImgProc/EdgeDetection.vb
Imports System.Drawing.Imaging
Imports System.Threading
Imports System.Runtime.InteropServices.Marshal

Public Class EdgeDetection : Inherits ComImgProc
    Public Sub New(_bitmap As Bitmap)
        MyBase.New(_bitmap)
    End Sub

    Protected Overrides Sub Finalize()
        MyBase.Finalize()
    End Sub

    Public Overrides Function GoImgProc(_token As CancellationToken) As Boolean
        Dim bRst As Boolean = True

        Dim nMask As Short(,) =
        {
            {1, 1, 1},
            {1, -8, 1},
            {1, 1, 1}
        }

        Dim nWidthSize As Integer = MyBase.m_bitmap.Width
        Dim nHeightSize As Integer = MyBase.m_bitmap.Height
        Dim nMasksize As Integer = nMask.GetLength(0)

        MyBase.m_bitmapAfter = New Bitmap(MyBase.m_bitmap)

        Dim bitmapData As BitmapData = MyBase.m_bitmapAfter.LockBits(New Rectangle(0, 0, nWidthSize, nHeightSize), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb)

        Dim nIdxWidth As Integer
        Dim nIdxHeight As Integer

        For nIdxHeight = 0 To nHeightSize - 1 Step 1
            For nIdxWidth = 0 To nWidthSize - 1 Step 1
                If (_token.IsCancellationRequested) Then
                    bRst = False
                    Exit For
                End If

                Dim pAdr As IntPtr = bitmapData.Scan0
                Dim nPos As Integer = nIdxHeight * bitmapData.Stride + nIdxWidth * 4
                Dim nPixelB As Byte = ReadByte(pAdr, nPos + ComInfo.Pixel.B)
                Dim nPixelG As Byte = ReadByte(pAdr, nPos + ComInfo.Pixel.G)
                Dim nPixelR As Byte = ReadByte(pAdr, nPos + ComInfo.Pixel.R)
                Dim nPixelA As Byte = ReadByte(pAdr, nPos + ComInfo.Pixel.A)

                Dim lCalB As Long = 0.0
                Dim lCalG As Long = 0.0
                Dim lCalR As Long = 0.0

                Dim nIdxWidthMask As Integer
                Dim nIdxHightMask As Integer

                For nIdxHightMask = 0 To nMasksize - 1 Step 1
                    For nIdxWidthMask = 0 To nMasksize - 1 Step 1
                        If (nIdxWidth + nIdxWidthMask > 0 And
                                nIdxWidth + nIdxWidthMask < nWidthSize And
                                nIdxHeight + nIdxHightMask > 0 And
                                nIdxHeight + nIdxHightMask < nHeightSize) Then

                            Dim pAdr2 As IntPtr = bitmapData.Scan0
                            Dim nPos2 As Integer = (nIdxHeight + nIdxHightMask) * bitmapData.Stride + (nIdxWidth + nIdxWidthMask) * 4

                            lCalB += ReadByte(pAdr2, nPos2 + ComInfo.Pixel.B) * nMask(nIdxWidthMask, nIdxHightMask)
                            lCalG += ReadByte(pAdr2, nPos2 + ComInfo.Pixel.G) * nMask(nIdxWidthMask, nIdxHightMask)
                            lCalR += ReadByte(pAdr2, nPos2 + ComInfo.Pixel.R) * nMask(nIdxWidthMask, nIdxHightMask)
                        End If
                    Next
                Next
                WriteByte(pAdr, nPos + ComInfo.Pixel.B, ComFunc.LongToByte(lCalB))
                WriteByte(pAdr, nPos + ComInfo.Pixel.G, ComFunc.LongToByte(lCalG))
                WriteByte(pAdr, nPos + ComInfo.Pixel.R, ComFunc.LongToByte(lCalR))
            Next
        Next
        MyBase.m_bitmapAfter.UnlockBits(bitmapData)

        Return bRst
    End Function
End Class